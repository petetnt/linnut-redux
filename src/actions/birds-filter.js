export const FORM_VALUE_CHANGE = 'FORM_VALUE_CHANGE';
export const ACTIVE_MARKER_CHANGE = 'ACTIVE_MARKER_CHANGE';
export const FETCH_FAMILIES = 'FETCH_FAMILIES';

export const formValueChange = (value, fieldName) => ({
  type: FORM_VALUE_CHANGE,
  value,
  fieldName,
});
export const activeMarkerChange = id => ({
  type: ACTIVE_MARKER_CHANGE,
  id,
});

// export const fetchFamilies = () => {
//   return dispatch => {
//     Birds.on('value', snapshot => {
//       dispatch({
//         type: FETCH_POSTS,
//         payload: snapshot.val(),
//       });
//     });
//   };
// }
