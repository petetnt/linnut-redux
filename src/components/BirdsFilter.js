import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const FiltersContainer = styled.section`
  position: absolute;
  width: 100%;
  height: auto;
  background-color: #fff;
  padding: 1em;
  top: 0;
  left: 0;
  z-index: 100;
`;

class BirdsFilter extends React.Component {
  render() {
    const {
      filters,
      onChange,
    } = this.props;

    return (
      <FiltersContainer>
        { filters.map(filter => (

          <select
            key={filter.id}
            ref={filter.id}
            name={filter.id}
            onChange={onChange}
            value={filter.value || ''}
          >
            <option value="">{ filter.id }</option>
            { filter.options.map((option, index) => (
              <option
                // eslint-disable-next-line react/no-array-index-key
                key={index}
                value={option}
              >
                { index } - { option }
              </option>),
            )}
          </select>),
        )}
      </FiltersContainer>
    );
  }
}
BirdsFilter.propTypes = {
  onChange: PropTypes.func.isRequired,
  filters: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string,
    value: PropTypes.string,
    options: PropTypes.arrayOf(PropTypes.string),
  })).isRequired,
  filteredBirds: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number,
    PrimaryName: PropTypes.string,
    SecondaryName: PropTypes.string,
    Date: PropTypes.string,
    Count: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    PlaceName: PropTypes.string,
    Latitude: PropTypes.number,
    Longitude: PropTypes.number,
    Notes: PropTypes.string,
    Family: PropTypes.string,
    FamilyName: PropTypes.string,
  })).isRequired,
};
export default BirdsFilter;
