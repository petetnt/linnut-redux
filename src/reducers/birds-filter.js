import birds from '../data/birds';
import helpers from '../helpers';
import {
  FORM_VALUE_CHANGE,
} from '../actions/birds-filter';

const Families = [...new Set(birds.map(bird => bird.Family))];
const PrimaryNames = [...new Set(birds.map(bird => bird.PrimaryName))];
const dates = [...new Set(birds.map(bird => helpers.getYear(bird.Date)))];

const filters = [
  { id: 'Family', value: '', options: Families },
  { id: 'Date', value: '', options: dates },
  { id: 'PrimaryName', value: '', options: PrimaryNames },
];

const primaryNameOptionsByFamilyValue = (initialBirds, currentFamily) =>
  [...new Set(initialBirds.filter(bird =>
    bird.Family === currentFamily)
  .map(bird => bird.PrimaryName))];

const filterBirds = (initialBirds, nextFilters) => (
  nextFilters.reduce((acc, filter) => acc.filter((elem) => {
    if (filter.value === '') {
      return elem;
    }
    if (filter.id === 'Date') {
      return filter.value === helpers.getYear(elem[filter.id]);
    }
    return filter.value === elem[filter.id];
  }), initialBirds)
);

const initialState = {
  birds,
  filters,
  filteredBirds: birds,
};

const birdsFilter = (state = initialState, action) => {
  switch (action.type) {
  case FORM_VALUE_CHANGE: {
    const id = action.fieldName;

    const nextFilters = state.filters.map((filter) => {
      /* Updates options for PrimaryName dropdown if Family is changed */
      if (filter.id === 'PrimaryName' && id === 'Family') {
        const nextPrimaryNames =
          primaryNameOptionsByFamilyValue(state.birds, action.value);
        return { ...filter, options: nextPrimaryNames, value: '' };
      }
      if (filter.id === id) {
        return { ...filter, value: action.value };
      }
      return filter;
    });

    const nextBirds = filterBirds(state.birds, nextFilters);

    return {
      ...state,
      filteredBirds: nextBirds,
      filters: nextFilters,
    };
  }
  default:
    return state;
  }
};

export default birdsFilter;
