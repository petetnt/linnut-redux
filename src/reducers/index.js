import { combineReducers } from 'redux';
import { firebaseStateReducer } from 'react-redux-firebase';
import sightingsFilter from './sightings-filter';

const rootReducer = combineReducers({
  birds: sightingsFilter,
  firebase: firebaseStateReducer,
});

export default rootReducer;
