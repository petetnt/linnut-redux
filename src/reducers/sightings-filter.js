import h from '../helpers';
import moment from 'moment';
import {
  ACTIVE_MARKER_CHANGE,
  FORM_VALUE_CHANGE,
  FETCH_INITIAL_DATA,
} from '../actions/sightings-filter';

const filters = [
  { id: 'familyName', value: '', displayName: '', options: [] },
  { id: 'Year', value: '', displayName: '', options: [] },
  { id: 'birdName', value: '', displayName: '', options: [] },
];
const mapConfig = {
  center: {
    lat: 60.1634549,
    lng: 24.9449212,
  },
};
const initialState = {
  sightings: [],
  filters,
  filteredSightings: [],
  mapConfig,
  activeMarker: '',
};

const getBirdNames = (sightings, family, year) => {
  // console.log([
  //   ...new Set(
  //     Object.keys(sightings)
  //       .filter(key => {
  //         return
  //         // if (!family) {
  //         //   console.log('tuli tänne');
  //         //   return h.getYear(sightings[key].Date) === year;
  //         // } else if (!year) {
  //         //   console.log('tuli tänne myös');
  //         //   return sightings[key].Family === family;
  //         // }
  //         // return (
  //         //   sightings[key].Family === family &&
  //         //   h.getYear(sightings[key].Date) === year
  //         // );
  //       })
  //       .map(bird => bird.PrimaryName),
  //   ),
  // ]);
  return [
    ...new Set(
      Object.keys(sightings).map(key => {
        return sightings[key].bird.name;
      }),
    ),
  ];
};

const getUniqueProperties = (sightings, property) => {
  return [
    ...new Set(
      Object.keys(sightings).map(key => {
        return sightings[key].bird[property];
      }),
    ),
  ];
};
const sightingsFilter = (state = initialState, action) => {
  switch (action.type) {
    case ACTIVE_MARKER_CHANGE: {
      return {
        ...state,
        activeMarker: action.id,
      };
    }
    case FORM_VALUE_CHANGE: {
      return {
        ...state,
        activeMarker: 2,
      };
    }
    case FETCH_INITIAL_DATA: {
      const sightings = action.payload;
      const newFamilies = getUniqueProperties(sightings, 'familyName');
      const newBirds = getUniqueProperties(sightings, 'name');
      console.log(newFamilies);
      console.log(newBirds);

      const nextFilters = filters.map(filter => {
        /* Updates options for PrimaryName dropdown if Family or Date is changed */
        // const oldFamily = filters.find(f => f.id === 'Family').value;
        // const oldYear = filters.find(f => f.id === 'Date').value;
        // const currentFamily = id === 'Family' ? action.value : oldFamily;
        // const currentDate = id === 'Date' ? action.value : oldDate;

        if (filter.id === 'birdName') {
          const nextBirdNames = getBirdNames(sightings, '', '');
          return {
            ...filter,
            options: nextBirdNames,
            value: '',
            displayName: '',
          };
        } else {
          return { ...filter };
        }
        return filter;
      });

      // const nextsightings = filtersightings(state.sightings, nextFilters);

      return {
        ...state,
        filteredSightings: sightings,
        filters: nextFilters,
        sightings: action.payload,
      };
    }
    default:
      return state;
  }
};

// const Families = [...new Set(sightings.map(bird => bird.Family))];
// const PrimaryNames = [...new Set(sightings.map(bird => bird.PrimaryName))];
// const dates = [...new Set(sightings.map(bird => helpers.getYear(bird.Date)))];
// const activeMarker = {
//   id: '',
// };
// const mapConfig = {
//   center: {
//     lat: 60.1634549,
//     lng: 24.9449212,
//   },
// };
// export const filters = [
//   { id: 'Family', value: '', displayName: '', options: Families },
//   { id: 'Date', value: '', displayName: '', options: dates },
//   { id: 'PrimaryName', value: '', displayName: '', options: PrimaryNames },
// ];

// const updatePrimaryNames = (initialsightings, currentFamily, currentDate) => [
//   ...new Set(
//     initialsightings
//       .filter(bird => {
//         if (!currentFamily) {
//           return helpers.getYear(bird.Date) === currentDate;
//         } else if (!currentDate) {
//           return bird.Family === currentFamily;
//         }
//         return (
//           bird.Family === currentFamily &&
//           helpers.getYear(bird.Date) === currentDate
//         );
//       })
//       .map(bird => bird.PrimaryName),
//   ),
// ];

// const filtersightings = (initialsightings, nextFilters) =>
//   nextFilters.reduce(
//     (acc, filter) =>
//       acc.filter(elem => {
//         if (filter.value === '') {
//           return elem;
//         }
//         if (filter.id === 'Date') {
//           return filter.value === helpers.getYear(elem[filter.id]);
//         }
//         return filter.value === elem[filter.id];
//       }),
//     initialsightings,
//   );

// const initialState = {
//   sightings,
//   filters,
//   filteredsightings: sightings,
//   mapConfig,
//   activeMarker,
// };

// const sightingsFilter = (state = initialState, action) => {
//   switch (action.type) {
//     case ACTIVE_MARKER_CHANGE: {
//       return {
//         ...state,
//         activeMarker: action.id,
//       };
//     }
//     case FORM_VALUE_CHANGE: {
//       const id = action.fieldName;

//       const nextFilters = state.filters.map(filter => {
//         /* Updates options for PrimaryName dropdown if Family or Date is changed */
//         if (filter.id === 'PrimaryName' && (id === 'Family' || id === 'Date')) {
//           const oldFamily = state.filters.find(f => f.id === 'Family').value;
//           const oldDate = state.filters.find(f => f.id === 'Date').value;
//           const currentFamily = id === 'Family' ? action.value : oldFamily;
//           const currentDate = id === 'Date' ? action.value : oldDate;
//           const nextPrimaryNames = updatePrimaryNames(
//             state.sightings,
//             currentFamily,
//             currentDate,
//           );
//           return {
//             ...filter,
//             options: nextPrimaryNames,
//             value: '',
//             displayName: '',
//           };
//         }
//         if (filter.id === id) {
//           return { ...filter, value: action.value };
//         }
//         return filter;
//       });

//       const nextsightings = filtersightings(state.sightings, nextFilters);

//       return {
//         ...state,
//         filteredsightings: nextsightings,
//         filters: nextFilters,
//       };
//     }
//     default:
//       return state;
//   }
// };

export default sightingsFilter;
