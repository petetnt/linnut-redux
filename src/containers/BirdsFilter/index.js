import React from 'react';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { firebaseConnect } from 'react-redux-firebase';
import PropTypes from 'prop-types';
import BirdsFilter from '../../components/BirdsFilter';
import BirdsMap from '../BirdsMap';
import {
  formValueChange,
  activeMarkerChange,
  fetchSightings,
} from '../../actions/sightings-filter';

const BirdsFilterContainer = props => {
  const { onChange, onMarkerClick } = props;
  return (
    <div>
      {/* 
        <BirdsFilter {...props} onChange={onChange} />
        <BirdsMap {...props} onClick={onMarkerClick} />
      */}
    </div>
  );
};

function mapStateToProps(state) {
  const { families, sightings } = state.firebase.data;

  const { filters, filteredBirds, mapConfig, activeMarker } = state.birds;

  return {
    families,
    sightings,
    filteredBirds,
    filters,
    mapConfig,
    activeMarker,
  };
}

const mapDispatchToProps = dispatch => ({
  onChange: event => {
    dispatch(
      formValueChange(event.currentTarget.value, event.currentTarget.name),
    );
  },
  onMarkerClick: event => {
    dispatch(activeMarkerChange(event));
  },
});

BirdsFilterContainer.propTypes = {
  onChange: PropTypes.func.isRequired,
  onMarkerClick: PropTypes.func.isRequired,
};
export default compose(
  firebaseConnect(['/families', '/sightings']),
  connect(mapStateToProps, mapDispatchToProps),
)(BirdsFilterContainer);
