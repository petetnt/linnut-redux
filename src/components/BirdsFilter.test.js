import renderer from 'react-test-renderer';
import React from 'react';
import BirdsFilter from './BirdsFilter';
import { filters } from '../reducers/birds-filter';

const filteredBirds = [];
const onChange = () => {};

test('BirdsFilter renders correctly', () => {
  const tree = renderer.create(
    <BirdsFilter filters={filters} onChange={onChange} filteredBirds={filteredBirds} />,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
