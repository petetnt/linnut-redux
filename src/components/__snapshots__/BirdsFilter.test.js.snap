exports[`test BirdsFilter renders correctly 1`] = `
<div>
  <select
    name="Family"
    onChange={[Function]}
    value="">
    <option
      value="">
      Family
    </option>
    <option
      value="tikat">
      tikat
    </option>
    <option
      value="varikset">
      varikset
    </option>
    <option
      value="kahlaajat">
      kahlaajat
    </option>
    <option
      value="lokit">
      lokit
    </option>
    <option
      value="rastaat">
      rastaat
    </option>
    <option
      value="tiaiset">
      tiaiset
    </option>
    <option
      value="peipot">
      peipot
    </option>
    <option
      value="sorsat">
      sorsat
    </option>
    <option
      value="joutsenet-hanhet">
      joutsenet-hanhet
    </option>
    <option
      value="varpuset-kiurut-kirviset">
      varpuset-kiurut-kirviset
    </option>
    <option
      value="kurki-haikarat">
      kurki-haikarat
    </option>
    <option
      value="kyyhkyt">
      kyyhkyt
    </option>
    <option
      value="petolinnut">
      petolinnut
    </option>
    <option
      value="yolaulajat">
      yolaulajat
    </option>
    <option
      value="vastarakit-taskut-lepinkaiset">
      vastarakit-taskut-lepinkaiset
    </option>
    <option
      value="uikut-kuikat">
      uikut-kuikat
    </option>
    <option
      value="kertut">
      kertut
    </option>
    <option
      value="paaskyt">
      paaskyt
    </option>
  </select>
  <select
    name="Date"
    onChange={[Function]}
    value="">
    <option
      value="">
      Date
    </option>
    <option
      value="2016">
      2016
    </option>
  </select>
  <select
    name="PrimaryName"
    onChange={[Function]}
    value="">
    <option
      value="">
      PrimaryName
    </option>
    <option
      value="palokärki">
      palokärki
    </option>
    <option
      value="harakka">
      harakka
    </option>
    <option
      value="töyhtöhyyppä">
      töyhtöhyyppä
    </option>
    <option
      value="naurulokki">
      naurulokki
    </option>
    <option
      value="mustarastas">
      mustarastas
    </option>
    <option
      value="talitiainen">
      talitiainen
    </option>
    <option
      value="sinitiainen">
      sinitiainen
    </option>
    <option
      value="Naakka">
      Naakka
    </option>
    <option
      value="Varis">
      Varis
    </option>
    <option
      value="kottarainen">
      kottarainen
    </option>
    <option
      value="peippo">
      peippo
    </option>
    <option
      value="Uivelo">
      Uivelo
    </option>
    <option
      value="telkkä">
      telkkä
    </option>
    <option
      value="haapana">
      haapana
    </option>
    <option
      value="laulujoutsen">
      laulujoutsen
    </option>
    <option
      value="nokikana">
      nokikana
    </option>
    <option
      value="sinisorsa">
      sinisorsa
    </option>
    <option
      value="tikli">
      tikli
    </option>
    <option
      value="varpunen">
      varpunen
    </option>
    <option
      value="kalalokki">
      kalalokki
    </option>
    <option
      value="punarinta">
      punarinta
    </option>
    <option
      value="kurki">
      kurki
    </option>
    <option
      value="räkättirastas">
      räkättirastas
    </option>
    <option
      value="valkoposkihanhi">
      valkoposkihanhi
    </option>
    <option
      value="kuovi">
      kuovi
    </option>
    <option
      value="kyhmyjoutsen">
      kyhmyjoutsen
    </option>
    <option
      value="meriharakka">
      meriharakka
    </option>
    <option
      value="sepelkyyhky">
      sepelkyyhky
    </option>
    <option
      value="haahka">
      haahka
    </option>
    <option
      value="viherpeippo">
      viherpeippo
    </option>
    <option
      value="kalatiira">
      kalatiira
    </option>
    <option
      value="ruskosuohaukka">
      ruskosuohaukka
    </option>
    <option
      value="satakieli">
      satakieli
    </option>
    <option
      value="kivitasku">
      kivitasku
    </option>
    <option
      value="kirjosieppo">
      kirjosieppo
    </option>
    <option
      value="kuikka">
      kuikka
    </option>
    <option
      value="tukkasotka">
      tukkasotka
    </option>
    <option
      value="silkkiuikku">
      silkkiuikku
    </option>
    <option
      value="Isokoskelo">
      Isokoskelo
    </option>
    <option
      value="mustapääkerttu">
      mustapääkerttu
    </option>
    <option
      value="kapustarinta">
      kapustarinta
    </option>
    <option
      value="keltavästäräkki">
      keltavästäräkki
    </option>
    <option
      value="tervapääsky">
      tervapääsky
    </option>
    <option
      value="räystäspääsky">
      räystäspääsky
    </option>
    <option
      value="harmaasorsa">
      harmaasorsa
    </option>
    <option
      value="korppi">
      korppi
    </option>
    <option
      value="pikkuvarpunen">
      pikkuvarpunen
    </option>
    <option
      value="kuusitiainen">
      kuusitiainen
    </option>
    <option
      value="tilhi">
      tilhi
    </option>
    <option
      value="punatulkku">
      punatulkku
    </option>
    <option
      value="pähkinähakki">
      pähkinähakki
    </option>
    <option
      value="närhi">
      närhi
    </option>
    <option
      value="Urpiainen">
      Urpiainen
    </option>
    <option
      value="harmaapäätikka">
      harmaapäätikka
    </option>
  </select>
</div>
`;
